lib/mocha.js

```diff
@@ -1108,7 +1271,40 @@ Mocha.prototype.lazyLoadFiles = function lazyLoadFiles(enable) {
 
 /**
  * An function that returns a {@link MochaRootHookObject}, either sync or async.
- * @private
- * @callback MochaRootHookFunction
+   @callback MochaRootHookFunction
  * @returns {MochaRootHookObject|Promise<MochaRootHookObject>}
  */
+
+/**
+ * A function that's invoked _once_ which is either sync or async.
+ * Can be a "teardown" or "setup".  These will all share the same context.
+ * @callback MochaGlobalFixture
+ * @returns {void|Promise<void>}
+ */
+
+/**
```

Comment:
This describes the shape of a plugin definition. the builtin plugins are defined in `lib/plugin-loader.js` around L24 (`MochaPlugins`).  A plugin definition defines everything about how a user would provide their plugin implementation, and how that implementation ends up in a `Mocha` instance.  once the implementation reaches the `Mocha` instance, it's no longer managed by this definition.  It's _not_ atomic in this way--it's tightly coupled to the `Mocha` class, which must do whatever it needs to do to "use" the implementation.  e.g., the plugin definition says "use the `mochaHooks` export, it should pass this validation function, and call the resulting option `rootHooks` when instantiating `Mocha`".  after that, it's `Mocha`'s problem.

what this looks like exactly is _not_ something we need to make public atm. I tried to keep the abstraction minimal, but it may not be enough or it may be too much. in other words, we don't support 3rd-party plugin definitions

Reply:
I think it would be _nice_ to make it completely self-contained, but we just don't have any place to "hook in" to Mocha to handle the implementation (e.g., "run any global setup fixtures as async functions--in the main process only--before the `run` event is emitted").  adding those hooks would be _very_ involved, and this PR is big enough already.